File: 5회차-linearDS/5회차 필수문제 코드/CPP/2812.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, k;
    cin >> n >> k;

    string s;
    cin >> s;

    deque<int> stk;
    int cnt = 0;
    for (int i = 0; i < (int)s.length(); ++i) {
        int now = s[i] - '0';
        if (stk.empty() || stk.back() >= now)
            stk.push_back(now);
        else {
            while (!stk.empty() && stk.back() < now && cnt < k) {
                stk.pop_back();
                cnt++;
            }
            stk.push_back(now);
        }
    }
    
    while (cnt < k) {
        stk.pop_back();
        cnt++;
    }

    for (auto it : stk)
        cout << it;
    cout << '\n';
}

File: 5회차-linearDS/5회차 필수문제 코드/PYTHON/2812.py
n, k = map(int, input().split())
s = input()

stk = []
cnt = 0
for i in range(len(s)):
    now = int(s[i])
    if len(stk) == 0 or stk[-1] >= now:
        stk.append(now)
    else:
        while len(stk) != 0 and stk[-1] < now and cnt < k:
            stk.pop()
            cnt += 1
        stk.append(now)

while cnt < k:
    stk.pop()
    cnt += 1

for it in stk:
    print(it , end="")